File: templates/Uwp/_comp/CodeBehind/Feat.IdentityCommon._VB/Services/UserDataService.vb
Imports System.Collections.Generic
Imports System.Threading.Tasks
Imports Param_RootNamespace.Core.Helpers
Imports Param_RootNamespace.Core.Models
Imports Param_RootNamespace.Core.Services
Imports Param_RootNamespace.Helpers
Imports Param_RootNamespace.Models
Imports Windows.Storage

Namespace Services
    Public Class UserDataService
        Private Const _userSettingsKey As String = "IdentityUser"
        Private _user As UserData

        Private ReadOnly Property IdentityService As IdentityService
            Get
                Return Singleton(Of IdentityService).Instance
            End Get
        End Property

        Private ReadOnly Property MicrosoftGraphService As MicrosoftGraphService
            Get
                Return Singleton(Of MicrosoftGraphService).Instance
            End Get
        End Property

        Public Event UserDataUpdated As EventHandler(Of UserData)

        Public Sub New()
        End Sub

        Public Sub Initialize()
            AddHandler IdentityService.LoggedIn, AddressOf OnLoggedIn
            AddHandler IdentityService.LoggedOut, AddressOf OnLoggedOut
        End Sub

        Public Async Function GetUserAsync() As Task(Of UserData)
            If _user Is Nothing Then
                _user = Await GetUserFromCacheAsync()

                If _user Is Nothing Then
                    _user = GetDefaultUserData()
                End If
            End If

            Return _user
        End Function

        Private Async Sub OnLoggedIn(sender As Object, e As EventArgs)
            _user = Await GetUserFromGraphApiAsync()
            RaiseEvent UserDataUpdated(Me, _user)
        End Sub

        Private Async Sub OnLoggedOut(sender As Object, e As EventArgs)
            _user = Nothing
            Await ApplicationData.Current.LocalFolder.SaveAsync(Of User)(_userSettingsKey, Nothing)
        End Sub

        Private Async Function GetUserFromCacheAsync() As Task(Of UserData)
            Dim cacheData = Await ApplicationData.Current.LocalFolder.ReadAsync(Of User)(_userSettingsKey)
            Return Await GetUserDataFromModel(cacheData)
        End Function

        Private Async Function GetUserFromGraphApiAsync() As Task(Of UserData)
            Dim accessToken = Await IdentityService.GetAccessTokenAsync()

            If String.IsNullOrEmpty(accessToken) Then
                Return Nothing
            End If

            Dim userData = Await MicrosoftGraphService.GetUserInfoAsync(accessToken)

            If userData IsNot Nothing Then
                userData.Photo = Await MicrosoftGraphService.GetUserPhoto(accessToken)
                Await ApplicationData.Current.LocalFolder.SaveAsync(_userSettingsKey, userData)
            End If

            Return Await GetUserDataFromModel(userData)
        End Function

        Private Async Function GetUserDataFromModel(userData As User) As Task(Of UserData)
            If userData Is Nothing Then
                Return Nothing
            End If

            Dim userPhoto = If(String.IsNullOrEmpty(userData.Photo), ImageHelper.ImageFromAssetsFile("DefaultIcon.png"), Await ImageHelper.ImageFromStringAsync(userData.Photo))
            Return New UserData() With {
                .Name = userData.DisplayName,
                .UserPrincipalName = userData.UserPrincipalName,
                .Photo = userPhoto
            }
        End Function

        Private Function GetDefaultUserData() As UserData
            Return New UserData() With {
                .Name = IdentityService.GetAccountUserName(),
                .Photo = ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
            }
        End Function
    End Class
End Namespace

File: code/TemplateStudioForUWP/Templates/_comp/CB/IdentityCommon/Services/UserDataService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Param_RootNamespace.Core.Helpers;
using Param_RootNamespace.Core.Models;
using Param_RootNamespace.Core.Services;
using Param_RootNamespace.Helpers;
using Param_RootNamespace.Models;
using Windows.Storage;

namespace Param_RootNamespace.Services
{
    public class UserDataService
    {
        private const string _userSettingsKey = "IdentityUser";

        private UserData _user;

        private IdentityService IdentityService => Singleton<IdentityService>.Instance;

        private MicrosoftGraphService MicrosoftGraphService => Singleton<MicrosoftGraphService>.Instance;

        public event EventHandler<UserData> UserDataUpdated;

        public UserDataService()
        {
        }

        public void Initialize()
        {
            IdentityService.LoggedIn += OnLoggedIn;
            IdentityService.LoggedOut += OnLoggedOut;
        }

        public async Task<UserData> GetUserAsync()
        {
            if (_user == null)
            {
                _user = await GetUserFromCacheAsync();
                if (_user == null)
                {
                    _user = GetDefaultUserData();
                }
            }

            return _user;
        }

        private async void OnLoggedIn(object sender, EventArgs e)
        {
            _user = await GetUserFromGraphApiAsync();
            UserDataUpdated?.Invoke(this, _user);
        }

        private async void OnLoggedOut(object sender, EventArgs e)
        {
            _user = null;
            await ApplicationData.Current.LocalFolder.SaveAsync<User>(_userSettingsKey, null);
        }

        private async Task<UserData> GetUserFromCacheAsync()
        {
            var cacheData = await ApplicationData.Current.LocalFolder.ReadAsync<User>(_userSettingsKey);
            return await GetUserDataFromModel(cacheData);
        }

        private async Task<UserData> GetUserFromGraphApiAsync()
        {
            var accessToken = await IdentityService.GetAccessTokenForGraphAsync();
            if (string.IsNullOrEmpty(accessToken))
            {
                return null;
            }

            var userData = await MicrosoftGraphService.GetUserInfoAsync(accessToken);
            if (userData != null)
            {
                userData.Photo = await MicrosoftGraphService.GetUserPhoto(accessToken);
                await ApplicationData.Current.LocalFolder.SaveAsync(_userSettingsKey, userData);
            }

            return await GetUserDataFromModel(userData);
        }

        private async Task<UserData> GetUserDataFromModel(User userData)
        {
            if (userData == null)
            {
                return null;
            }

            var userPhoto = string.IsNullOrEmpty(userData.Photo)
                ? ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
                : await ImageHelper.ImageFromStringAsync(userData.Photo);

            return new UserData()
            {
                Name = userData.DisplayName,
                UserPrincipalName = userData.UserPrincipalName,
                Photo = userPhoto
            };
        }

        private UserData GetDefaultUserData()
        {
            return new UserData()
            {
                Name = IdentityService.GetAccountUserName(),
                Photo = ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
            };
        }
    }
}
